File: surface/widgets/winui/components/widgets/primitives/FocusTrap.cs
// ============================================================
// Clef Surface WinUI Widget — FocusTrap
//
// Focus-trap wrapper that constrains focus within a boundary.
// When active, focus is captured within the content scope;
// when inactive, children render normally.
//
// Adapts the focus-trap.widget spec: anatomy (root,
// sentinelStart, sentinelEnd), states (inactive, active), and
// connect attributes to WinUI 3 rendering.
// ============================================================

using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Input;

namespace Clef.Surface.WinUI.Widgets.Primitives;

public sealed class ClefFocusTrap : UserControl
{
    public static readonly DependencyProperty IsActiveProperty =
        DependencyProperty.Register(nameof(IsActive), typeof(bool), typeof(ClefFocusTrap),
            new PropertyMetadata(false, OnPropertyChanged));

    public static readonly DependencyProperty ShouldLoopProperty =
        DependencyProperty.Register(nameof(ShouldLoop), typeof(bool), typeof(ClefFocusTrap),
            new PropertyMetadata(true));

    public bool IsActive { get => (bool)GetValue(IsActiveProperty); set => SetValue(IsActiveProperty, value); }
    public bool ShouldLoop { get => (bool)GetValue(ShouldLoopProperty); set => SetValue(ShouldLoopProperty, value); }

    private readonly ContentPresenter _contentPresenter;
    private readonly StackPanel _root;

    public ClefFocusTrap()
    {
        _contentPresenter = new ContentPresenter();
        _root = new StackPanel();
        _root.Children.Add(_contentPresenter);
        Content = _root;
        IsTabStop = true;
    }

    public void SetTrapContent(UIElement element)
    {
        _contentPresenter.Content = element;
    }

    private static void OnPropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
    {
        var trap = (ClefFocusTrap)d;
        if (trap.IsActive)
        {
            trap.Focus(FocusState.Programmatic);
        }
    }
}

File: surface/widgets/gtk/components/widgets/primitives/FocusTrap.ts
// ============================================================
// Clef Surface GTK Widget — FocusTrap
//
// Focus-constraining container that keeps keyboard focus within
// its child hierarchy. Uses GTK4 focus management to prevent
// tab-out when active.
//
// Adapts the focus-trap.widget spec: anatomy (root), states
// (active, inactive), and connect attributes to GTK4 rendering.
// ============================================================

import Gtk from 'gi://Gtk?version=4.0';

// --------------- Props ---------------

export interface FocusTrapProps {
  active?: boolean;
  child?: Gtk.Widget | null;
}

// --------------- Component ---------------

/**
 * Creates a GTK4 focus trap container that constrains keyboard
 * navigation within its child widget hierarchy.
 */
export function createFocusTrap(props: FocusTrapProps = {}): Gtk.Widget {
  const { active = true, child = null } = props;

  const frame = new Gtk.Box({
    orientation: Gtk.Orientation.VERTICAL,
  });

  if (active) {
    frame.set_focusable(true);
    frame.set_focus_on_click(true);
    frame.set_can_focus(true);
  }

  if (child) {
    frame.append(child);
  }

  return frame;
}
